src/nn/variation/architecture_adaptation.py

```diff
@@ -162,4 +204,172 @@ def adapt_node_shape(graph, node, current_size, target_size, target_user=None, t
+    Args:
+        graph: The FX graph
+        node: The node whose shape needs to be adapted
+        current_size: Current size of the node's output, no batch dimension
+        target_size: Desired size of the node's output, no batch dimension
+        target_user: Optional specific node that should use the adapted output. If None, all users will be updated.
+    Returns:
+        graph: The modified graph
+        adapted_node: The node after shape adaptation
+    """
+    # Get total elements of current_size and target_size
+    current_total = math.prod(current_size)
+    target_total = math.prod(target_size)
+    logging.debug(f"Shape adaptation: {current_size} -> {target_size} (total elements: {current_total} -> {target_total})")
+    
+    # Calculate GCF between current_total and target_total
+    gcf = _calculate_gcf(current_total, target_total)
+    reduced_current = current_total // gcf
+    reduced_target = target_total // gcf
+    logging.debug(f"Found GCF={gcf}, reducing problem from {current_total}:{target_total} to {reduced_current}:{reduced_target}")
+    
+    # Use a single reshape to (gcf, reduced_current)
+    graph, node = add_specific_node(
+        graph,
+        node,
+        ReshapeModule((gcf, reduced_current)),
+        target_user=target_user
+    )
+    logging.debug(f"Reshaped to factor out GCF: (batch, {gcf}, {reduced_current})")
+    
+    # Determine if we're upsampling or downsampling the reduced dimension
+    is_upsampling = reduced_target > reduced_current
+    logging.debug(f"Operation type: {'upsampling' if is_upsampling else 'downsampling'}")
+    
+    if is_upsampling:
+        # For upsampling, work with target/current ratio
+        feature_ratio = reduced_target / reduced_current
+        r1 = math.floor(feature_ratio)
+        r2 = math.ceil(feature_ratio)
+        if r1 != r2:
+            length_scale = int((reduced_target - (r2*reduced_current)) / (r1 - r2))
+        else:
+            length_scale = reduced_current
+        r1_slice_length = length_scale
+        r1_shape_values = (length_scale, 1, r1)
+        r2_slice_length = reduced_current - r1_slice_length
+        r2_shape_values = (r2_slice_length, 1, r2)
+        logging.debug(f"Upsampling ratios: r1={r1}, r2={r2}, scale={length_scale}")
+    else:
+        # For downsampling, work with current/target ratio
+        feature_ratio = reduced_current / reduced_target
+        r1 = math.floor(feature_ratio)
+        r2 = math.ceil(feature_ratio)
+        # Only process if r1 != r2
+        if r1 != r2:
+            length_scale = int((reduced_current - (r2*reduced_target)) / (r1 - r2))
+        else:
+            length_scale = reduced_current
+        r1_slice_length = r1 * length_scale
+        r1_shape_values = (length_scale, r1, 1)
+        r2_slice_length = reduced_current - r1_slice_length
+        r2_shape_values = (r2_slice_length//r2, r2, 1)
+        logging.debug(f"Downsampling ratios: r1={r1}, r2={r2}, scale={length_scale}")
```

Comment:
```suggestion
    if is_upsampling:  # For upsampling, work with target/current ratio
        feature_ratio = reduced_target / reduced_current
        smaller_num_elements = reduced_current
        larger_num_elements = reduced_target
    else:  # For downsampling, work with current/target ratio
        feature_ratio = reduced_current / reduced_target
        smaller_num_elements = reduced_target
        larger_num_elements = reduced_current
    

    feature_ratio = larger_num_elements / smaller_num_elements
    r1 = math.floor(feature_ratio)
    r2 = math.ceil(feature_ratio)
    # Only process if r1 != r2
    if r1 != r2:
        length_scale = int((larger_num_elements - (r2*smaller_num_elements)) / (r1 - r2))
    else:
        length_scale = reduced_current
    r1_slice_length = r1 * length_scale
    r1_shape_values = (length_scale, r1, 1)
    r2_slice_length = reduced_current - r1_slice_length
    r2_shape_values = (r2_slice_length//r2, r2, 1)
    logging.debug(f"ratios: r1={r1}, r2={r2}, scale={length_scale}")
```

Could we do it this way to reduce duplicate logic?

Reply:
r1 and r2 slice shape formulas are different for up and down sampling

r1 and r2 calculations are the only things that can be calculated once, check the latest commit